### PR TITLE
Add CSV Parser operator resolves #176 

### LINKED ISSUE #176: CSV Parser
This is just an optimization on regex parsing, but it would be pretty simple to support csv as a native parsing format.
```
type: csv_parser
format: first,last,phone,address,etc
```

### PR DESCRIPTION
**Pull request by ericwholt**
## Description of Changes
Adds CSV Parser operator and updates documentation.

Resolves https://github.com/observIQ/stanza/issues/176

## **Please check that the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Add a changelog entry (for non-trivial bug fixes / features)
- [x] CI passes

### CODE REVIEW

File: operator/builtin/parser/csv/csv.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	switch value.(type) {
+	case string:
+		csvLine += value.(string)
+	case []byte:
+		csvLine += string(value.([]byte))
+	default:
+		return nil, fmt.Errorf("type '%T' cannot be parsed as csv", value)
+	}
+
+  delimiterStr := string([]rune{r.fieldDelimiter})
```
Comment (djaglowski) [inline operator/builtin/parser/csv/csv.go:95]: Indentation error?

File: operator/builtin/parser/csv/csv_test.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  require.Contains(t, err.Error(), "record on line 1: wrong number of fields")
+}
+
+func TestCSVParserInvalidType(t *testing.T) {
+  parser := newTestParser(t)
+	_, err := parser.parse([]int{})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "type '[]int' cannot be parsed as csv")
+}
+
```
Comment (djaglowski) [inline operator/builtin/parser/csv/csv_test.go:52]: We should test some additional invalid inputs including:
- unparsed json
- wrong delimiter
- multiple lines (especially where they are each valid on their own)

File: operator/builtin/parser/csv/csv.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	record, err := reader.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	l := len(record)
+	if l != 1 {
+		return nil, fmt.Errorf("expected to parse a single csv record, got '%d'", l)
+	}
```
Comment (djaglowski) [inline operator/builtin/parser/csv/csv.go:110]: I don't think we should use `reader.ReadAll()` if we're just going to error if multiple lines are found. Could we possibly handle multiple records?

**Review by djaglowski — COMMENTED** [review]
This generally looks good. A couple comments and questions though.

File: operator/builtin/parser/csv/csv.go (hunk 3 above)
Comment (jsirianni) [reply operator/builtin/parser/csv/csv.go:110]: Our thinking was "we should never have more than one line per entry", however, which is correct for the current behavior of file input. I agree that it is a poor assumption though, we will rework it to support multiple lines in one entry.

Comment (djaglowski) [reply operator/builtin/parser/csv/csv.go:110]: It's probably a reasonable assumption for now. Maybe we should just use `reader.Read` though? Then we don't have to check the length?

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.7414087|+0.051710248|124.62769|+1.4352112|
|1|5000|6.2414765|+0.051548958|133.95663|+3.2175446|
|1|10000|12.551979|+0.94824314|138.28085|-3.1687622|
|1|50000|58.362926|-2.1202927|174.25027|-2.254303|
|1|100000|116.24186|+11.847801|225.97293|+0.1689148|
|10|100|2.5518212|+0.24137616|130.72858|+4.050911|
|10|500|6.05182|-1.7246175|135.04984|-0.31492615|
|10|1000|13.241573|-1.0002289|143.78098|+1.5022888|
|10|5000|68.56469|+4.4934387|180.1343|-7.187912|
|10|10000|124.897896|+7.154915|216.47185|+4.3401184|

**Comment by djaglowski** [comment]
| Log Files | Logs / Second | CPU Avg (%) | CPU Avg Δ (%) | Memory Avg (MB) | Memory Avg Δ (MB) |
| --- | --- | --- | --- | --- | --- |
|1|1000|1.6896803|-1.8119812e-05|123.81654|+0.6240616|
|1|5000|6.7243824|+0.5344548|132.18912|+1.4500275|
|1|10000|12.758947|+1.1552114|139.43845|-2.0111694|
|1|50000|60.59924|+0.1160202|176.79741|+0.29283142|
|1|100000|112.413246|+8.019188|222.48424|-3.3197784|
|10|100|2.344923|+0.03447795|128.64816|+1.9704971|
|10|500|6.637972|-1.1384654|135.16702|-0.19773865|
|10|1000|13.776183|-0.4656191|143.85223|+1.5735474|
|10|5000|65.16987|+1.0986176|180.92618|-6.3960266|
|10|10000|124.40496|+6.6619797|223.24232|+11.110596|

**Review by djaglowski — APPROVED** [review]
LGTM

**Review by jsirianni — APPROVED** [review]

**Pull request merged**